audio/out/ao_openal.c

```diff
@@ -89,42 +91,189 @@ static int control(struct ao *ao, enum aocontrol cmd, void *arg)
+                return alGetEnumValue((ALchar*)"AL_FORMAT_51CHN16");
+            }
+            break;
+        case 4:
+            if (alGetEnumValue((ALchar*)"AL_FORMAT_QUAD16")) {
+                return alGetEnumValue((ALchar*)"AL_FORMAT_QUAD16");
+            }
+            break;
+        case 2:
+            if (alGetEnumValue((ALchar*)"AL_FORMAT_STEREO16")) {
+                return alGetEnumValue((ALchar*)"AL_FORMAT_STEREO16");
+            }
+            break;
+        case 1:
+            if (alGetEnumValue((ALchar*)"AL_FORMAT_MONO16")) {
+                return alGetEnumValue((ALchar*)"AL_FORMAT_MONO16");
+            }
+            break;
+        }
+
+    case AF_FORMAT_S32:
+        if (strstr(alGetString(AL_RENDERER), "X-Fi") != NULL) {
+            switch (channels) {
+            case 8:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_71CHN32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_71CHN32");
+                }
+                break;
+            case 7:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_61CHN32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_61CHN32");
+                }
+                break;
+            case 6:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_51CHN32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_51CHN32");
+                }
+                break;
+            case 4:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_QUAD32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_QUAD32");
+                }
+                break;
+            case 2:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_STEREO32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_STEREO32");
+                }
+                break;
+            case 1:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_MONO32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_MONO32");
+                }
+                break;
+            }
+        }
+
+    case AF_FORMAT_FLOAT:
+        if (alIsExtensionPresent((ALchar*)"AL_EXT_float32") == AL_TRUE) {
+            switch (channels) {
+            case 8:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_71CHN32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_71CHN32");
+                }
+                break;
+            case 7:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_61CHN32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_61CHN32");
+                }
+                break;
+            case 6:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_51CHN32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_51CHN32");
+                }
+                break;
+            case 4:
+                if (alGetEnumValue((ALchar*)"AL_FORMAT_QUAD32")) {
+                    return alGetEnumValue((ALchar*)"AL_FORMAT_QUAD32");
```

Comment:
shouldn't this be `AL_FORMAT_QUAD_FLOAT32`?
likewise for the formats above

Reply:
Make sure to note your discoveries in a comment inside `get_supported_layout()`.

Reply:
Done, I hope.